### DISCUSSION TITLE
Autofixture+Xunit: How to inject XunitLogger<T> from ILogger<T>

### DISCUSSION THREAD

**Discussion by dilandau2001** [AutoFixture/AutoFixture#1369] [Q&A]
### Problem Description

In this project I am using Autofixture to automatically create classes under test with the minimum amount of code possible.
For that I have followed the quickstart to create the AutoDomainDataAttribute.
https://autofixture.github.io/docs/quick-start/#

You end up with the following code
![image](https://user-images.githubusercontent.com/12834469/204466634-6efc4481-2330-49ec-afb8-884ea742193f.png)

With that attribute you can create tests like the following:
```cs
    public class MyUnitTests
    {
        private readonly ITestOutputHelper _testOutputHelper;

        public MyUnitTests(ITestOutputHelper testOutputHelper)
        {
            _testOutputHelper = testOutputHelper;
        }

        [Theory]
        [AutoDomainData]
        public void WhenThen(MyTestClass sut)
        {
            // Arrange
            
            // Act
            var res = sut.Example();

            // Assert
            Assert.True(res);
        }
    }
```

Where MyTestClass is the following:

```cs
    public class MyTestClass
    {
        private readonly ILogger<MyTestClass> _logger;

        public MyTestClass(ILogger<MyTestClass> logger)
        {
            _logger = logger;
            _logger.BeginScope(nameof(MyTestClass));
        }

        public bool Example()
        {
            using (_logger.BeginScope(nameof(Example)))
            {
                _logger.LogDebug("My test.");
                return true;
            }
        }
    }
```

Note that this class needs a `ILogger<T>` to be constructed, and that is done by autofixture creating a `Mock<ILogger<T>>` automatically.

### Objective/Goal
My goal is somehow inject the Xunit _testOutputHelper so traces from the tested class appears in the test log.
So for the given code, test output would say
[DBG] MyTestClass => Example => My Test.

### Constraints

- `ILogger` comes from Microsoft Logging extensions and most of its members are extension methods. Which means that they cannot be mocked in a setup. In order to setup something you would need it to be strict behavior and setup over the base functions. See [Link](https://adamstorr.azurewebsites.net/blog/mocking-ilogger-with-moq)
- AutomoqData will give you mocked behaviour Default on mocked types.
- `Mock<T>` cannot change its behaviour once created.  [Link](https://github.com/moq/moq4/issues/1230)
- TestOutPutHelper is created by Xunit in a per test basis. And I found no way to programatically get to it in a factory given "a context". So right now I am evaluating by setting a static property although I am sure that won't work when pararell test put in practice.

### What I tried so far

0 - Make a static helper with inputs ILogger<T> and _testOutputHelper to setup all possible uses of the logger inside the tested class.
I can get to the ILogger<T> instance by using [Frozen]. This approach would have work fine if ILogger from Microsoft not having extension methods as it has.  You still will miss traces from inside the constructor and you cannot setup over extension methods, and you cannot change behaviour once the mock is created. (See constraints)

1- Customize the fixture so the mocks created are strict. See [Link](https://stackoverflow.com/questions/34055672/how-can-i-make-automoqcustomization-use-strict-mockbehavior). It fails on constructors that make use of the mocked type inside the constructor, because you had no time to setup the mock. And I don't know why it only applies strict to the first mocked parameter.

2- Try to customize the AutoDomainData to identify `ILogger<T>` to create a XunitLogger. Fixture doesn't support generic types.

3 - Create a Xunit Logger provider See [Link](https://www.meziantou.net/how-to-get-asp-net-core-logs-in-the-output-of-xunit-tests.htm). Yes, I have a provider, but how do I tell Autofixture to use my provider? if autofixture cannot recognize the generic type. Is there something like a container I could register things to so autofixture try to resolve constructors using it?

4 - Use Moq.Automocker. See [Link](https://github.com/moq/Moq.AutoMocker/issues/126). I lack the knowledge to add automocker to the Autofixture fixture pipeline. I feel this is similar to point 3. Not sure If this would work although I think it could. (except for the TestoutputHelper constraint which is not singleton)

### My current working way 

Although ugly my current approach is to make loggers public so I can set them inside the test. I am still missing the calls inside the constructor and I am overriding the injected instance.

```cs
    public class MyTestClass
    {   
        public MyTestClass(ILogger<MyTestClass> logger)
        {
            Logger = logger;
            Logger.BeginScope(nameof(MyTestClass));
        }

        public ILogger<MyTestClass> Logger { get; set; }

        public bool Example()
        {
            using (Logger.BeginScope(nameof(Example)))
            {
                Logger.LogDebug("My test");
                return true;
            }
        }
    }

    public class MyUnitTests
    {
        private readonly XUnitLoggerProvider _loggerProvider;
        
        public MyUnitTests(ITestOutputHelper testOutputHelper)
        {
            _loggerProvider = new XUnitLoggerProvider(testOutputHelper);
        }

        [Theory]
        [AutoDomainData]
        public void WhenThen(MyTestClass sut)
        {
            // Arrange
            sut.Logger = _loggerProvider.CreateLogger<MyTestClass>();
            
            // Act
            var res = sut.Example();

            // Assert
            Assert.True(res);
        }
    }
```

**Comment by aivascu** [comment]
@dilandau2001 unfortunately XUnit 2 does not expose a lot of integration points that we could use to capture the output helper instantiated by the test runner. The situation might change in XUnit 3, but until then this is impossible to do via the data attribute.

The only way to inject the custom logger, that I can think of at the moment, is to let XUnit inject the fixture in the test constructor.
You can do this by creating  a child `Fixture` class that has a single parameterless constructor and implement `IClassFixutre` in your test class.

You'll have to give up the data attribute, but you'll be able to see the logs in the Test Explorer window.

```cs
public class DomainFixture : Fixture
{
    public DomainFixture() : base()
    {
        this.OmitAutoProperties = true;
        this.Customize(new AutoMoqCustomization());
    }
}

public class UnitTest1 : IClassFixture<DomainFixture>
{
    private readonly ITestOutputHelper outputHelper;
    private readonly DomainFixture fixture;

    public UnitTest1(ITestOutputHelper outputHelper, DomainFixture fixture)
    {
        this.outputHelper = outputHelper;
        this.fixture = fixture;
        var provider = new XUnitLoggerProvider(outputHelper);
        this.fixture.Inject(provider.CreateLogger<MyTestClass>());
    }

    [Fact]
    public void WhenThen()
    {
        // Arrange
        var sut = fixture.Create<MyTestClass>();

        // Act
        var res = sut.Example();

        // Assert
        Assert.True(res);
    }
}
```
![image](https://user-images.githubusercontent.com/7030530/204671237-382179f3-6518-4d83-85ee-3df217f16aed.png)

**Comment by maxisam** [comment]
https://github.com/apazureck/AutoFixture.Community.LoggerFactoryCustomizations/blob/master/README.md this should simplfy the process a bit
